tests/submodule/lookup.c

```diff
@@ -388,3 +388,28 @@ void test_submodule_lookup__renamed(void)
 	cl_git_pass(git_submodule_foreach(g_repo, sm_lookup_cb, &data));
 	cl_assert_equal_i(8, data.count);
 }
+
+void test_submodule_lookup_cached(void) {
+	git_submodule *sm;
+	git_submodule *sm2;
+	/* See that the simple tests still pass. */
+
+	git_repository_submodule_cache_all(g_repo);
+	test_submodule_lookup__simple_lookup();
+	git_repository_submodule_cache_clear(g_repo);
+	test_submodule_lookup__simple_lookup();
+
+	/* Check that subsequent calls return different objects when cached. */
+	git_repository_submodule_cache_all(g_repo);
+	cl_assert(0 == git_submodule_lookup(&sm, g_repo, "sm_unchanged"));
+	cl_assert(0 == git_submodule_lookup(&sm2, g_repo, "sm_unchanged"));
+	cl_assert(sm == sm2);
```

Comment:
Using `cl_assert` is not ideal here, since it just will say "an assertion failed".  We have a variety of helper functions that provide more explicit messages, which are helpful in debugging and understanding test failures:

For git functions, please use `cl_git_pass`, which will show the error messages properly when a function fails and returns nonzero.  (eg, `cl_git_pass(git_submodule_lookup...)`)

To compare two pointers, please use `cl_assert_equal_p` which will show more information in a nice format about the actual and expected values.